crypto/params.c

```diff
@@ -0,0 +1,208 @@
+    }
+
+    memcpy(p->buffer, val, width);
+    return 1;
+}
+
+#define PARAM_INT(name, type, PARAM) \
+    int OSSL_PARAM_get_##name(const OSSL_PARAM *p, const char *key, type *val) \
+    { \
+        return OSSL_PARAM_get_int_common(p, key, val, PARAM); \
+    } \
+    int OSSL_PARAM_set_##name(const OSSL_PARAM *p, const char *key, type val) \
+    { \
+        return OSSL_PARAM_set_int_common(p, key, &val, PARAM); \
+    }
+
+PARAM_INT(int, int, OSSL_PARAM_INT)
+PARAM_INT(uint, unsigned int, OSSL_PARAM_UINT)
+PARAM_INT(int64, int64_t, OSSL_PARAM_INT64)
+PARAM_INT(uint64, uint64_t, OSSL_PARAM_UINT64)
+PARAM_INT(long, long int, OSSL_PARAM_LONG)
+PARAM_INT(ulong, unsigned long int, OSSL_PARAM_ULONG)
+PARAM_INT(size_t, size_t, OSSL_PARAM_SIZET)
+
+/*
+ * In the get/set double functions, "width" is always sizeof(double) but
+ * having parallel construction with the integer functions above seems
+ * worthwhile.
+ */
+
+int OSSL_PARAM_get_double(const OSSL_PARAM *p, const char *key, double *val)
+{
+    size_t width;
+    double *dp;
+
+
+    if ((p = OSSL_PARAM_locate(p, key)) == NULL) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_GET_DOUBLE, CRYPTO_R_PARAM_NOT_FOUND);
+        ERR_add_error_data(2, "param name=", key);
+        return 0;
+    }
+
+    /* Type safety. */
+    width = OSSL_PARAM_get_width(OSSL_PARAM_DOUBLE);
+    if (p->data_type != OSSL_PARAM_DOUBLE || p->buffer_size != width) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_GET_DOUBLE, CRYPTO_R_TYPE_MISMATCH);
+        return 0;
+    }
+
+    dp = (double *)p->buffer;
+    *val = *dp;
+    return 1;
+}
+
+int OSSL_PARAM_set_double(const OSSL_PARAM *p, const char *key, double val)
+{
+    size_t width;
+    double *dp;
+
+
+    if ((p = OSSL_PARAM_locate(p, key)) == NULL) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_SET_DOUBLE, CRYPTO_R_PARAM_NOT_FOUND);
+        ERR_add_error_data(2, "param name=", key);
+        return 0;
+    }
+
+    /* Type safety. */
+    width = OSSL_PARAM_get_width(OSSL_PARAM_DOUBLE);
+    if (p->return_size != NULL)
+        *p->return_size = width;
+    if (p->data_type != OSSL_PARAM_DOUBLE || p->buffer_size != width) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_SET_DOUBLE, CRYPTO_R_TYPE_MISMATCH);
+        return 0;
+    }
+    dp = (double *)p->buffer;
+    *dp = val;
+    return 1;
+}
+
+int OSSL_PARAM_get_bignum(const OSSL_PARAM *p, const char *key, BIGNUM **val)
+{
+    BIGNUM *b;
+
+
+    if ((p = OSSL_PARAM_locate(p, key)) == NULL) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_GET_BIGNUM, CRYPTO_R_PARAM_NOT_FOUND);
+        ERR_add_error_data(2, "param name=", key);
+        return 0;
+    }
+
+    /* Type safety. */
+    if (p->data_type != OSSL_PARAM_BIGNUM) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_GET_BIGNUM, CRYPTO_R_TYPE_MISMATCH);
+        return 0;
+    }
+
+    if ((b = BN_native2bn(p->buffer, (int)p->bn_size, *val)) == NULL)
+        return 0;
+    *val = b;
+    return 1;
+}
+
+int OSSL_PARAM_set_bignum(OSSL_PARAM *p, const char *key, const BIGNUM *val)
```

Comment:
There's no reason why `p` shouldn't be const here. Undo the `bn_size` stuff and you will be safe.
Have a look in my "EXPERIMENTAL" test to see how I solved it.